### PR TITLE
ALL: Unify zlib and gzio API

### PR DESCRIPTION
**Pull request by lephilousophe**
All functions are made the same and existing ones in gzio are changed to look like our historical functions.

zlib is used when available and falls back on gzio.
This allows performance improvements as our CRC32 and gzio implementations are slower than base zlib.
As zlib is available when libpng is present, this is sensible to benefit from it.

A simple benchmark using valgrind shows a 3x speed up in term of cycles between zlib and gzio, zlib being the faster.

This PR should make #4514 obsolete.

Only ScummVM startup has been tested for now hence the draft status.

### CODE REVIEW

File: common/compression/unzip.cpp [hunk 1]
```diff
@@ -953,8 +969,11 @@ Common::SharedArchiveContents unzOpenCurrentFile (unzFile file, const Common::CR
 		delete[] compressedBuffer;
 		return Common::SharedArchiveContents();
 	}
-
+#ifndef USE_ZLIB
 	uint32 crc32_data = crc.crcFast(uncompressedBuffer, s->cur_file_info.uncompressed_size);
+#else
+	uint32 crc32_data = crc32(0, uncompressedBuffer, s->cur_file_info.uncompressed_size);
```
Comment (ccawley2011) [inline common/compression/unzip.cpp:975]: Is it worth having a common wrapper for this as well?

File: common/compression/vise.cpp
Comment (ccawley2011) [inline common/compression/vise.cpp:1]: There's a leftover `USE_ZLIB` ifdef in this file.

**Comment by ccawley2011** [comment]
The grim, hdb, icb, sword25, titanic and wintermute engines have zlib as dependencies which can probably be removed now.

File: common/compression/unzip.cpp (hunk 1 above)
Comment (lephilousophe) [reply common/compression/unzip.cpp:975]: I don't see easily how to fix this.
Maybe in gzio we could create a crc32 function which has a static CRC object?

File: common/compression/vise.cpp
Comment (lephilousophe) [reply common/compression/vise.cpp:1]: Thanks, fixed.

**Comment by lephilousophe** [comment]
> The grim, hdb, icb, sword25, titanic and wintermute engines have zlib as dependencies which can probably be removed now.

I didn't even know there was such dependency! Thanks!

File: engines/director/util.cpp [hunk 2]
```diff
@@ -1136,22 +1136,18 @@ uint32 readVarInt(Common::SeekableReadStream &stream) {
 }
 
 Common::SeekableReadStreamEndian *readZlibData(Common::SeekableReadStream &stream, unsigned long len, unsigned long *outLen, bool bigEndian) {
-#ifdef USE_ZLIB
 	byte *in = (byte *)malloc(len);
 	byte *out = (byte *)malloc(*outLen);
 	stream.read(in, len);
 
-	if (!Common::uncompress(out, outLen, in, len)) {
+	if (!Common::inflateZlib(out, outLen, in, len)) {
```
Comment (sev-) [inline engines/director/util.cpp:1143]: Hmm, isn't it `deflateZlib`?

**Comment by sev-** [comment]
The vast majority of changes look pretty straightforward to me. LGTM.

**Review by phcoder — COMMENTED** [review]
I like the direction. I have one comment. I didn't yet read the full patch. How do you handle click team case?

**Comment by lephilousophe** [comment]
> I like the direction. I have one comment. I didn't yet read the full patch. How do you handle click team case?
I still keep gzio around for Clickteam case. zlib is used for all other standard cases only.

This PR has now been tested with code coverage to ensure the majority of the code is correct in both modes (with and without zlib):
- `inflateZlib` using some dcr Shockwave file with director engine,
- `inflateZlibHeaderless` using zip parser (GUI) and with `inflateZlibInstallShield` (more on this later)
- `inflateClickteam` using Maniac Mansion Deluxe (mmdsetup.exe),
- `wrapCompressedReadStream` using a GZiped save game,
- `wrapDeflateReadStream` **only in gzio** through `inflateZlibHeaderless`,
- `wrapClickteamReadStream` through `inflateClickteam`,
- `wrapCompressedWriteStream` **not tested** but not changed either,
- `inflateZlibInstallShield` replacement using CAB from Feeble Files (with markers) and Nancy 3 (without markers)
- `inflateZlibHeaderless` replacement using Cumberbund (Glk/ADRIFT).

For me, it's ready to go.

**Comment by sev-** [comment]
Great! Merging

**Pull request merged**